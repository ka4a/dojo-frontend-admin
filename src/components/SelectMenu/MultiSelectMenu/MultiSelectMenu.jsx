import { memo, useCallback, useState } from 'react';
import PropTypes from 'prop-types';
import { Dropdown } from '@edx/paragon';
import * as R from 'ramda';
import { MultiselectItem } from './components/MultiSelectItem/MultiSelectItem';

export const MultiselectMenu = memo(
  ({
    title, onChange, values, options, shouldChangeOnClose,
  }) => {
    const [localValueSets, setLocalValuesSets] = useState(new Set(values));
    const [isOpened, setIsOpened] = useState(false);

    const handleChange = useCallback(
      (changedValues) => {
        if (!R.equals(values, changedValues)) {
          onChange(changedValues);
        }
      },
      [values, localValueSets, onChange],
    );

    const handleClose = useCallback(() => {
      if (shouldChangeOnClose) {
        handleChange([...localValueSets]);
      }

      setIsOpened(false);
    }, [shouldChangeOnClose, localValueSets, handleChange]);

    const handleOpen = useCallback(() => {
      setIsOpened(true);

      if (!isOpened) {
        setLocalValuesSets(new Set(values));
      }
    }, [setIsOpened, isOpened, setLocalValuesSets, values]);

    const handleSelect = useCallback(
      (value) => {
        let selectedValues = [...localValueSets];

        if (!selectedValues.includes(value)) {
          selectedValues.push(value);
        } else {
          selectedValues = selectedValues.filter(
            (localValueSet) => localValueSet !== value,
          );
        }

        if (!shouldChangeOnClose) {
          handleChange(selectedValues);
        }

        setLocalValuesSets(new Set(selectedValues));
      },
      [
        values,
        localValueSets,
        setLocalValuesSets,
        shouldChangeOnClose,
        handleChange,
      ],
    );

    // todo: Paragon Dropdown onToggle doesn't work on component is opening. onClick is used to process the opening.

    return (
      <Dropdown autoClose="outside" onToggle={handleClose} onClick={handleOpen}>
        <Dropdown.Toggle variant="outline-extra" id="dropdown-basic-4">
          {title}
        </Dropdown.Toggle>
        <Dropdown.Menu>
          {options.map(({ name, value }) => (
            <MultiselectItem
              key={value}
              value={value}
              name={name}
              isSelected={localValueSets.has(value)}
              onSelect={handleSelect}
            />
          ))}
        </Dropdown.Menu>
      </Dropdown>
    );
  },
);

MultiselectMenu.propTypes = {
  title: PropTypes.string,
  options: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string,
      value: PropTypes.string,
    }),
  ),
  onChange: PropTypes.func,
  shouldChangeOnClose: PropTypes.bool,
  values: PropTypes.arrayOf(PropTypes.string),
};

MultiselectMenu.defaultProps = {
  title: '',
  options: [],
  values: [],
  onChange: null,
  shouldChangeOnClose: false,
};
