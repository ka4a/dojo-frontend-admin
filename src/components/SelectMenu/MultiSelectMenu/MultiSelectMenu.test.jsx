import { shallow, mount } from 'enzyme';
import { MultiselectMenu } from './MultiSelectMenu';

const defaultValues = {
  title: 'title',
  onChange: () => {},
  values: ['value'],
  options: [{ name: 'Name', value: 'value' }],
  shouldChangeOnClose: true,
};

describe('MultiSelectMenu', () => {
  it('should render MultiSelectMenu component', () => {
    shallow(<MultiselectMenu {...defaultValues} />);
  });

  it('it should render the expected HTML', () => {
    expect(
      mount(<MultiselectMenu {...defaultValues} />).html(),
    ).toMatchSnapshot();
  });
});
