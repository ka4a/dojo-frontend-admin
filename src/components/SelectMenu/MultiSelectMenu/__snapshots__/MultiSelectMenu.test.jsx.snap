// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiSelectMenu it should render the expected HTML 1`] = `"<div data-testid=\\"dropdown\\" class=\\"dropdown\\"><button aria-haspopup=\\"true\\" aria-expanded=\\"false\\" id=\\"dropdown-basic-4\\" type=\\"button\\" class=\\"dropdown-toggle btn btn-outline-extra\\">title</button></div>"`;
